### PR TITLE
Simplify PipelineTracer

### PR DESCRIPTION
**Pull request by jasnell**
Eliminate internal nesting of PipelineTracers. Use the resolved on complete promise to propagate traces up to the parent.

This was originally done as part of the larger PR to introduce streaming traces but makes sense to pull out separately here.

### CODE REVIEW

**Review by harrishancock — APPROVED** [review]

File: src/workerd/io/trace.h [hunk 1]
```diff
@@ -368,12 +361,12 @@ class PipelineTracer final: public kj::Refcounted {
   // to the host where tracing was initiated.
   void addTrace(rpc::Trace::Reader reader);
 
+  void addTracesFromChild(kj::ArrayPtr<kj::Own<Trace>> traces);
```
Comment (anonrig) [inline src/workerd/io/trace.h:364]: a documentation to this function would be nice

File: src/workerd/io/trace.c++ [hunk 2]
```diff
@@ -590,15 +591,15 @@ kj::Rc<WorkerTracer> PipelineTracer::makeWorkerTracer(PipelineLogLevel pipelineL
...
+  return kj::rc<WorkerTracer>(addRefToThis(), kj::mv(trace), pipelineLogLevel);
 }
 
 void PipelineTracer::addTrace(rpc::Trace::Reader reader) {
   traces.add(kj::refcounted<Trace>(reader));
 }
 
 WorkerTracer::WorkerTracer(
-    kj::Own<PipelineTracer> parentPipeline, kj::Own<Trace> trace, PipelineLogLevel pipelineLogLevel)
+    kj::Rc<PipelineTracer> parentPipeline, kj::Own<Trace> trace, PipelineLogLevel pipelineLogLevel)
```
Comment (anonrig) [inline src/workerd/io/trace.c++:602]: What's the reason for this change?

**Review by anonrig — APPROVED** [review]

File: src/workerd/io/trace.c++ (hunk 2 above)
Comment (jasnell) [reply src/workerd/io/trace.c++:602]: `kj::Rc<T>` is a newer alternative to `kj::Own<T>` that makes the refcounted semantics clearer (and safer). We haven't adopted it much yet but it definitely beats passing around bare references. Since I'm in here modifying this code anyway it makes sense to update to this.

File: src/workerd/io/trace.h (hunk 1 above)
Comment (jasnell) [inline src/workerd/io/trace.h:364]:
```suggestion
  // When collecting traces from multiple stages in a pipeline, this is called by the
  // tracer for a subordinate stage to add its collected traces to the parent pipeline.
  void addTracesFromChild(kj::ArrayPtr<kj::Own<Trace>> traces);
```
